### PR TITLE
Implement DynamicQuantizeLinear

### PR DESCRIPTION
**Pull request by acrrd**
Add `DynamicQuantizeLinear` operation to core and use it in onnx.

### CODE REVIEW

File: core/src/ops/quant.rs [hunk 1]
```diff
@@ -146,6 +147,119 @@ fn info_quantize_linear_i8(q: &QuantizeLinearI8) -> TractResult<Vec<String>> {
...
+    let min = v_min.min(0.);
+    let max = v_max.max(0.);
+
+    // quantize range
+    let min_t = 0.;
+    let max_t = 255.0;
+
+    let scale = (max - min) / max_t;
+
+    let zero_point = -min / scale;
```
Comment (janpetschexain) [inline core/src/ops/quant.rs:252]: this will panic if the input array contains zeros everywhere, i.e. `max == min`. this is quite unlikely in the real world, but could easily happen in test cases. we should probably catch this case with
```suggestion
    let (scale, zero_point) = if max == min {
        (1., 0.)
    } else {
        let scale = (max - min) / max_t;
        let zero_point = -min / scale;
        (scale, zero_point)
    };
```
for the edge case, scale can of course take any value except zero, but one is always a good choice for numerical stability when it is used in the division in `dynamic_quantize_linear_f32_u8`.

and we should probably also add a test for that edge case.

File: core/src/ops/quant.rs [hunk 2]
```diff
@@ -146,6 +147,119 @@ fn info_quantize_linear_i8(q: &QuantizeLinearI8) -> TractResult<Vec<String>> {
...
+        .max(u8::min_value() as i32)
+        .min(u8::max_value() as i32) as u8
+}
+
+fn dynamic_quantize_linear_u8(
+    scale: f32,
+    zero_point: u8,
+    xs: &[f32],
+    ys: &mut [u8],
+) -> TractResult<()> {
```
Comment (janpetschexain) [inline core/src/ops/quant.rs:161]: i guess this return value is required by convention of the tract lib? otherwise we could just get rid of it.

File: core/src/ops/quant.rs [hunk 3]
```diff
@@ -146,6 +147,119 @@ fn info_quantize_linear_i8(q: &QuantizeLinearI8) -> TractResult<Vec<String>> {
     )])
 }
 
+fn dynamic_quantize_linear_f32_u8(x: f32, scale: f32, zero_point: u8) -> u8 {
+    (((x / scale).round() as i32) + zero_point as i32)
+        .max(u8::min_value() as i32)
+        .min(u8::max_value() as i32) as u8
```
Comment (janpetschexain) [inline core/src/ops/quant.rs:153]: why not simply hard code this?
```suggestion
        .max(0.)
        .min(255.) as u8
```

File: core/src/ops/quant.rs (hunk 1 above)
Comment (acrrd) [reply core/src/ops/quant.rs:252]: The division is between `f32` and that is well defined also when `scale` is 0. The method `round` is also well defined when the value is `inf` or `Nan`.

File: core/src/ops/quant.rs (hunk 3 above)
Comment (acrrd) [reply core/src/ops/quant.rs:153]: I don't like magic numbers in the code. `u8::min_value()` is more explicit. We could use `u8::MIN` but the rest of the code is using `min_value`.

Comment (janpetschexain) [reply core/src/ops/quant.rs:153]: ok, makes sense.

File: core/src/ops/quant.rs (hunk 1 above)
Comment (janpetschexain) [reply core/src/ops/quant.rs]: for my records, it is `NaN as u8` which evaluates to `0` which makes this work.

**Review by janpetschexain — APPROVED** [review]
